src/core/appdir.cpp

```
@@ -656,7 +657,8 @@ namespace linuxdeploy {
             AppDir::AppDir(const std::string& path) : AppDir(fs::path(path)) {}
 
             void AppDir::setExcludeLibraryPatterns(const std::vector<std::string> &excludeLibraryPatterns) {
-                d->excludeLibraryPatterns = excludeLibraryPatterns;
+                d->excludeLibraryPatterns.insert(d->excludeLibraryPatterns.end(), excludeLibraryPatterns.begin(), excludeLibraryPatterns.end());
+                util::misc::stringVectorToEnv("LINUXDEPLOY_EXCLUDED_LIBRARIES", ';', d->excludeLibraryPatterns);
```

This line does not make sense. We should not influence the environment from a non-singleton class. As a user, you'd set the environment variable anyway, thus causing linuxdeploy-plugin-qt and subsequent runs of linuxdeploy from the script to pick up the environment variable. This is in line with all the other environment variables.

---

This is the whole reason of this patch.
user call ./ld --exclude-libraries foo --plugin qt
if linux deploy does not set the ENV var the qt plugin won't skipp foo library.

---

I makes no sense for
main ./linux-deploy --exclude-libraries foo works but for it's plugin the user has to manually set the env var.
Also there's no drawback from having duplicate items on the list, else we would use std::set and still doesn't mater for performance.

---

@TheAssassin ping

---

We do this for _all_ other environment variables, though. If you want the values to apply to all the plugins, you use the env var, otherwise, you use the parameter.

---

Then you should remove the cmd line option to exclude libraries, as it make no sense for it do not extend to it's plugins, I understand that a ENV var is the way to change called plugins, but this isn't specific for Qt plugin for example it should be honored by all plugins, plus there is no ENV var that plugins read to exclude-libraries.

---

Hey @dantti, I also ran into this (but with the gtk plugin while working on a PR for Tauri), and from my reading, I believe @TheAssassin (correct me if I'm wrong) just wants you to remove line 661. I think it's just a consistency thing. The CLI params only apply to linuxdeploy. Env vars apply to linuxdeploy + plugins. Propagating the CLI params here is just inconsistent with how linuxdeploy normally behaves. They don't want to propagate cli parameters and want to configure plugins via environment variables, per the spec: https://github.com/linuxdeploy/linuxdeploy/wiki/Plugin-system#controlling-plugin-behavior

> plus there is no ENV var that plugins read to exclude-libraries.

This PR to add $LINUXDEPLOY_EXCLUDED_LIBRARIES to linuxdeploy will be sufficient to add it to all plugins. The bash script based plugins will automatically work since they just bundle extra files and recall the original linuxdeploy package. linuxdeploy-plugin-qt will need to have it's linuxdeploy submodule bumped after this lands, but that's it.

---

@DistractionRectangle I could change to remove the cli to ENV passing, but that's just bad UX, I call `./ld --exclude=foo` and I expect that foo is going to be excluded, but it's not because plugins will still pick it, so instead I need to `LD_EXCLUDE=foo ./ld` and now I have it the way I wanted before, this to me is just bad UX. And users will still fill bug reports because `--exclude` didn't exclude...
I'd then say let's just remove all CLI options, otherwise you need to know how the magic works.

---

This just boils down to "why would you ever want to exclude a library from linuxdeploy and not from it's plugins?",  if the user if giving this CLI flag it's explicit that it should be excluded, if that's bad then it's the users's QA problem.

---

I agree with you about the UX, but that's probably a separate issue/PR. It seems to be the blocker for getting this merged, and getting this merged certainly improves the UX vs the workarounds we have to do now. 

Passing all cli options to plugins requires all plugins support them, or silently ignore invalid options, which hurts the UX and causes more overhead for the maintainer. It's also a 180 on the maintainers current stance, which is IMO a big ask and out of scope.